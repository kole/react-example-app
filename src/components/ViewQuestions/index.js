import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { inject, observer } from 'mobx-react';

import Category from '~/components/Category';
import QuestionBox from '~/components/QuestionBox';
import QuizProgress from '~/components/QuizProgress';
import AnswerButtons from '~/components/AnswerButtons';
import Spinner from '~/components/Spinner';

const QuestionsView = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    height: 100%;
    padding: 4em;
    text-align: center;

    > * {
        flex: 1;
        justify-content: center;
    }
`;

@inject('rootStore')
@observer
export default class Questions extends React.Component {
    componentDidMount() {
        const { rootStore: { QuestionsStore: { fetching, questions }}, history } = this.props;

        // If we are not fetching and we also don't have questions in the array,
        // send user back to Intro screen - there is a problem
        if (!fetching && questions.length === 0) {
            return history.push('/');
        }
    }

    getCurrentQuestionByNumber(num) {
        if (typeof num !== 'number') return;
        const { rootStore: { QuestionsStore: { questions, getQuestionAtIndex }}} = this.props;
        const idx = num - 1;
        if (questions.length === 0 || questions.length === idx) return;
        return getQuestionAtIndex(idx);
    }

    render() {
        const { rootStore: { QuestionsStore: { fetching, questions }}, match, history} = this.props;
        const currentPageNumber = parseInt(match.params.num, 10);
        const question = this.getCurrentQuestionByNumber(currentPageNumber);

        if (fetching || !question) return <Spinner />;

        return (
            <QuestionsView>
                <Category categoryText={question.category} />
                <QuestionBox questionText={question.question} />
                <QuizProgress currentQuestion={currentPageNumber} totalQuestions={questions.length} />
                <AnswerButtons history={history} questionID={currentPageNumber} />
            </QuestionsView>
        );
    }
}

Questions.propTypes = {
    rootStore: PropTypes.object,
    match: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired
}
