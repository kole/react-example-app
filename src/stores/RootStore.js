import Questions from './questions';
import Answers from './answers';
import Results from './results';

export default class RootStoreClass {
    constructor() {
        this.QuestionsStore = new Questions(this);
        this.AnswersStore = new Answers(this);
        this.ResultsStore = new Results(this);
    }
}
