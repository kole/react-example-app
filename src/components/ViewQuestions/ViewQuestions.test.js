/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import ViewQuestions from './index.js';
import RootStore from '~/stores/RootStore';
const store = new RootStore();

const mocks = { match: {params: {num: 1}}, history: {push: () => {}}};

describe('ViewQuestions', () => {
    test('Snapshot', () => {
        const component = renderer
            .create(<ViewQuestions history={mocks.history} match={mocks.match} rootStore={store} />)
            .toJSON();
        expect(component).toMatchSnapshot();
    });
});
