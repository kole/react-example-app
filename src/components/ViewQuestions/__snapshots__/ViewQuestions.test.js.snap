// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewQuestions Snapshot 1`] = `
<div
  className="sc-bZQynM fksxHI"
>
  <div>
    Loading
  </div>
  <svg
    height="44"
    stroke="#777"
    viewBox="0 0 44 44"
    width="44"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="none"
      fillRule="evenodd"
      strokeWidth="2"
    >
      <circle
        cx="22"
        cy="22"
        r="1"
      >
        <animate
          attributeName="r"
          begin="0s"
          calcMode="spline"
          dur="1.8s"
          keySplines="0.165, 0.84, 0.44, 1"
          keyTimes="0; 1"
          repeatCount="indefinite"
          values="1; 20"
        />
        <animate
          attributeName="stroke-opacity"
          begin="0s"
          calcMode="spline"
          dur="1.8s"
          keySplines="0.3, 0.61, 0.355, 1"
          keyTimes="0; 1"
          repeatCount="indefinite"
          values="1; 0"
        />
      </circle>
      <circle
        cx="22"
        cy="22"
        r="1"
      >
        <animate
          attributeName="r"
          begin="-0.9s"
          calcMode="spline"
          dur="1.8s"
          keySplines="0.165, 0.84, 0.44, 1"
          keyTimes="0; 1"
          repeatCount="indefinite"
          values="1; 20"
        />
        <animate
          attributeName="stroke-opacity"
          begin="-0.9s"
          calcMode="spline"
          dur="1.8s"
          keySplines="0.3, 0.61, 0.355, 1"
          keyTimes="0; 1"
          repeatCount="indefinite"
          values="1; 0"
        />
      </circle>
    </g>
  </svg>
</div>
`;
